README.md
# embedded-svm40 [![CircleCI](https://circleci.com/gh/Sensirion/embedded-svm40.svg?style=shield)](https://circleci.com/gh/Sensirion/embedded-svm40) [![GitHub license](https://img.shields.io/badge/license-BSD3-blue.svg)](https://raw.githubusercontent.com/Sensirion/embedded-svm40/master/LICENSE)

Embedded I2C driver for the Sensirion SVM40 module which measures humidity,
temperature and VOC in the air.

Compared to the SVM30 (which is included in the
https://github.com/Sensirion/embedded-sgp repository), the SVM40 combines
multiple sensors with a separate micro controller and is thus quite different
to use. For that reason the SVM40 driver lives in this separate repository.

## Download the Latest Driver Release

**Download the latest ready-to-use driver bundle from the [releases
page](https://github.com/Sensirion/embedded-svm40/releases/)**

We strongly recommend to use the driver from the release bundle instead of
cloning the repository.

## Upgrade the Firmware of the SVM40 Module

Firmware versions before 2.2 contain a bug where the sensor will return
constant measurement values after a few hours of operation (see
https://github.com/Sensirion/embedded-svm40/issues/21).

Use [Sensirion ControlCenter](https://www.sensirion.com/en/controlcenter/) to
upgrade the firmware of your device.

## Clone this repository
```
 git clone --recursive https://github.com/Sensirion/embedded-svm40.git
```

## Repository content
* `embedded-common` submodule repository for the common embedded driver HAL
* `svm40` SVM40 driver

## Collecting resources
```
make release
```
This will create the `release` folder with the necessary driver files in it,
including a Makefile. That way, you have just ONE folder with all the sources
ready to build your driver for your platform.

## Files to adjust (from embedded-common)
You only need to touch the following files:

* `sensirion_arch_config.h` (architecture specifics, you need to specify
  the integer sizes)

and depending on your i2c implementation either of the following:

* `embedded-common/hw_i2c/sensirion_hw_i2c_implementation.c`
  functions for hardware i2c communication if your platform supports that
* `embedded-common/sw_i2c/sensirion_sw_i2c_implementation.c`
  functions for software i2c communication via GPIOs

## Building the driver
1. Adjust `sensirion_arch_config.h` if you don't have the `<stdint.h>` header
   file available
2. Implement necessary functions in one of the `*_implementation.c` files
   described above
3. make

---

Please check the [embedded-common](https://github.com/Sensirion/embedded-common)
repository for further information and sample implementations.

---

## Other Resources

 * https://www.sensirion.com/en/environmental-sensors/gas-sensors/sgp40/ for
   more information about SGP40 and the SVM40 evaluation module
 * https://pypi.org/project/sensirion-shdlc-svm40/ for a Python driver for the
   SVM40 module.


CHANGELOG.md
# Changelog

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.3] - 2021-01-06

 * `[changed]` Print warning if the firmware version is older than 2.2 in the examples

## [0.2.2] - 2020-12-15

 * `[changed]` Updated embedded-common to 0.1.0 to improve compatibility when
               using multiple embedded drivers

## [0.2.1] - 2020-11-05

 * `[added]` Arduino Release zip and Documentation on how to run the included
             example usage on an Arduino platform.

## [0.2.0] - 2020-11-02

 * `[added]`   Reset Device function
 * `[added]`   Get serial command to SVM40 driver
 * `[added]`   Read Measured Values As Integers With Raw Parameters command to
               SVM40 driver
 * `[changed]` Add delays to all functions to match the hardware specification
 * `[changed]` Only include needed files from embedded-common in release zip

## [0.1.0] - 2020-09-01

 * First release with support for the basic commands

[Unreleased]: https://github.com/Sensirion/embedded-svm40/compare/0.2.3...master
[0.2.3]: https://github.com/Sensirion/embedded-svm40/compare/0.2.2...0.2.3
[0.2.2]: https://github.com/Sensirion/embedded-svm40/compare/0.2.1...0.2.2
[0.2.1]: https://github.com/Sensirion/embedded-svm40/compare/0.2.0...0.2.1
[0.2.0]: https://github.com/Sensirion/embedded-svm40/compare/0.1.0...0.2.0
[0.1.0]: https://github.com/Sensirion/embedded-svm40/releases/tag/0.1.0


tests/svm40-test.cpp
#include "sensirion_test_setup.h"
#include "svm40.h"
#include <inttypes.h>
#include <stdio.h>

#define MAX_VOC_INDEX 500
#define MIN_VOC_INDEX 0
#define MAX_VOC_RAW 65535
#define MIN_VOC_RAW 0
#define MAX_RH_RAW (119 * 100)
#define MIN_RH_RAW (-6 * 100)
#define MAX_T (130 * 200)
#define MIN_T (-45 * 200)

TEST_GROUP (SVM40_Tests) {
    void setup() {
        int16_t error;
        sensirion_i2c_init();

        // Select MUX 2 channel 7 (SVM40) on ch-sta-p-rp0002
        error = sensirion_i2c_mux_set_single_channel(0x72, 7);

        CHECK_ZERO_TEXT(error, "sensirion_i2c_mux_set_single_channel");
    }

    void teardown() {
        int16_t error;
        error = svm40_device_reset();
        CHECK_ZERO_TEXT(error, "svm40_device_reset");
        sensirion_i2c_release();
    }
};

TEST (SVM40_Tests, SVM40_Test_Driver_Version) {
    const char* version = svm40_get_driver_version();
    CHECK_TEXT(version, "svm40_get_driver_version();")
    printf("SVM40 Driver version: %s\n", version);
}

TEST (SVM40_Tests, SVM40_Test_Probe) {
    int16_t error;
    error = svm40_probe();
    CHECK_ZERO_TEXT(error, "svm40_probe");
}

TEST (SVM40_Tests, SVM40_Test_Get_Serial) {
    int16_t error;
    char serial_id[SVM40_MAX_SERIAL_LEN];
    error = svm40_get_serial(serial_id);
    CHECK_ZERO_TEXT(error, "svm40_get_serial");
    printf("Serial Number: %s\n", serial_id);
}

TEST (SVM40_Tests, SVM40_Test_Get_Version) {
    int16_t error;
    struct svm40_version_information version_information;
    error = svm40_get_version(&version_information);
    CHECK_ZERO_TEXT(error, "svm40_get_version");
    if (version_information.firmware_debug) {
        printf("Development firmware version: ");
    }
    printf(
        "FW: %u.%u, HW: %u.%u, protocol: %u.%u\n",
        version_information.firmware_major, version_information.firmware_minor,
        version_information.hardware_major, version_information.hardware_minor,
        version_information.protocol_major, version_information.protocol_minor);
}

TEST (SVM40_Tests, SVM40_Test_Measurement) {
    int16_t error;
    error = svm40_start_continuous_measurement();
    CHECK_ZERO_TEXT(error, "svm40_start_continuous_measurement");

    int16_t voc_index;
    int16_t relative_humidity;
    int16_t temperature;

    sensirion_sleep_usec(SVM40_MEASUREMENT_INTERVAL_USEC); /* wait 1s */
    error = svm40_read_measured_values_as_integers(
        &voc_index, &relative_humidity, &temperature);
    CHECK_ZERO_TEXT(error, "svm40_read_measured_values_as_integers");

    CHECK_TRUE_TEXT(voc_index >= MIN_VOC_INDEX && voc_index <= MAX_VOC_INDEX,
                    "svm40_read_measured_values_as_integers");
    CHECK_TRUE_TEXT(relative_humidity >= MIN_RH_RAW &&
                        relative_humidity <= MAX_RH_RAW,
                    "svm40_read_measured_values_as_integers");
    CHECK_TRUE_TEXT(temperature >= MIN_T && temperature <= MAX_T,
                    "svm40_read_measured_values_as_integers");

    error = svm40_stop_measurement();
    CHECK_ZERO_TEXT(error, "svm40_stop_measurement");
}

TEST (SVM40_Tests, SVM40_Test_Mesurement_Raw) {
    int16_t error;
    error = svm40_start_continuous_measurement();
    CHECK_ZERO_TEXT(error, "svm40_start_continuous_measurement");

    int16_t voc_index;
    int16_t relative_humidity;
    int16_t temperature;
    uint16_t voc_ticks_raw;
    int16_t uncompensated_relative_humidity_raw;
    int16_t uncompensated_temperature;

    sensirion_sleep_usec(SVM40_MEASUREMENT_INTERVAL_USEC); /* wait 1s */
    error = svm40_read_measured_values_as_integers_with_raw_params(
        &voc_index, &relative_humidity, &temperature, &voc_ticks_raw,
        &uncompensated_relative_humidity_raw, &uncompensated_temperature);
    CHECK_ZERO_TEXT(error,
                    "svm40_read_measured_values_as_integers_with_raw_params");

    CHECK_TRUE_TEXT(voc_index >= MIN_VOC_INDEX && voc_index <= MAX_VOC_INDEX,
                    "svm40_read_measured_values_as_integers");
    CHECK_TRUE_TEXT(relative_humidity >= MIN_RH_RAW &&
                        relative_humidity <= MAX_RH_RAW,
                    "svm40_read_measured_values_as_integers_with_raw_params");
    CHECK_TRUE_TEXT(temperature >= MIN_T && temperature <= MAX_T,
                    "svm40_read_measured_values_as_integers_with_raw_params");
    CHECK_TRUE_TEXT(uncompensated_relative_humidity_raw >= MIN_RH_RAW &&
                        uncompensated_relative_humidity_raw <= MAX_RH_RAW,
                    "svm40_read_measured_values_as_integers_with_raw_params");
    CHECK_TRUE_TEXT(uncompensated_temperature >= MIN_T &&
                        uncompensated_temperature <= MAX_T,
                    "svm40_read_measured_values_as_integers_with_raw_params");
    CHECK_TRUE_TEXT(voc_ticks_raw >= MIN_VOC_RAW &&
                        voc_ticks_raw <= MAX_VOC_RAW,
                    "svm40_read_measured_values_as_integers_with_raw_params");

    error = svm40_stop_measurement();
    CHECK_ZERO_TEXT(error, "svm40_stop_measurement");
}

TEST (SVM40_Tests, SVM40_Test_Reset_Device_Idle) {
    int16_t error;
    error = svm40_device_reset();
    CHECK_ZERO_TEXT(error, "svm40_device_reset");
}

TEST (SVM40_Tests, SVM40_Test_Reset_Device_Measure_Mode) {
    int16_t error;
    error = svm40_start_continuous_measurement();
    CHECK_ZERO_TEXT(error, "svm40_start_continuous_measurement");

    error = svm40_device_reset();
    CHECK_ZERO_TEXT(error, "svm40_device_reset");
}


svm40/svm40.c
/*
 * Copyright (c) 2020, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "svm40.h"
#include "svm40_git_version.h"

const char* svm40_get_driver_version(void) {
    return SVM40_DRV_VERSION_STR;
}

int16_t svm40_get_serial(char* serial) {
    int16_t error;

    error =
        sensirion_i2c_write_cmd(SVM40_I2C_ADDRESS, SVM40_CMD_GET_SERIAL_NUMBER);
    if (error != NO_ERROR) {
        return error;
    }
    sensirion_sleep_usec(SVM40_READ_DELAY);
    error = sensirion_i2c_read_words_as_bytes(
        SVM40_I2C_ADDRESS, (uint8_t*)serial,
        SVM40_MAX_SERIAL_LEN / SENSIRION_WORD_SIZE);
    if (error != NO_ERROR) {
        return error;
    }
    serial[SVM40_MAX_SERIAL_LEN - 1] = '\0';
    return NO_ERROR;
}

int16_t svm40_probe(void) {
    struct svm40_version_information version_information;
    return svm40_get_version(&version_information);
}

int16_t svm40_start_continuous_measurement(void) {
    return sensirion_i2c_write_cmd(SVM40_I2C_ADDRESS,
                                   SVM40_CMD_START_CONTINUOUS_MEASUREMENT);
}

int16_t svm40_stop_measurement(void) {
    return sensirion_i2c_write_cmd(SVM40_I2C_ADDRESS,
                                   SVM40_CMD_STOP_MEASUREMENT);
}

int16_t svm40_read_measured_values_as_integers(int16_t* voc_index,
                                               int16_t* relative_humidity,
                                               int16_t* temperature) {
    int16_t error;
    int16_t buffer[3];
    error = sensirion_i2c_write_cmd(SVM40_I2C_ADDRESS,
                                    SVM40_CMD_READ_MEASURED_VALUES_AS_INTEGERS);
    if (error) {
        return error;
    }

    error = sensirion_i2c_read_words(SVM40_I2C_ADDRESS, (uint16_t*)buffer,
                                     SENSIRION_NUM_WORDS(buffer));
    if (error) {
        return error;
    }

    *voc_index = buffer[0];
    *relative_humidity = buffer[1];
    *temperature = buffer[2];

    return NO_ERROR;
}

int16_t
svm40_get_version(struct svm40_version_information* version_information) {
    int16_t error;
    uint8_t buffer[8];
    error = sensirion_i2c_write_cmd(SVM40_I2C_ADDRESS, SVM40_CMD_GET_VERSION);
    if (error) {
        return error;
    }

    error = sensirion_i2c_read_words_as_bytes(SVM40_I2C_ADDRESS, buffer,
                                              SENSIRION_NUM_WORDS(buffer));

    if (error) {
        return error;
    }

    version_information->firmware_major = buffer[0];
    version_information->firmware_minor = buffer[1];
    version_information->firmware_debug = buffer[2];
    version_information->hardware_major = buffer[3];
    version_information->hardware_minor = buffer[4];
    version_information->protocol_major = buffer[5];
    version_information->protocol_minor = buffer[6];
    // the last byte is a padding byte

    return NO_ERROR;
}


svm40/svm40_example_usage.c
/*
 * Copyright (c) 2020, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "svm40.h"

#include <stdio.h>

/* TO USE CONSOLE OUTPUT (printf) YOU MAY NEED TO ADAPT THE
 * INCLUDES ABOVE OR DEFINE THEM ACCORDING TO YOUR PLATFORM.
 * #define printf(...)
 */

int main(void) {
    int16_t error;
    char serial_id[SVM40_MAX_SERIAL_LEN];

    const char* driver_version = svm40_get_driver_version();
    if (driver_version) {
        printf("SVM40 driver version %s\n", driver_version);
    } else {
        printf("fatal: Getting driver version failed\n");
        return -1;
    }

    /* Initialize I2C bus */
    sensirion_i2c_init();

    while (svm40_probe() != 0) {
        printf("SVM40 sensor probing failed\n");
        sensirion_sleep_usec(1000000); /* wait 1s */
    }
    printf("SVM40 sensor probing successful\n");

    error = svm40_get_serial(serial_id);
    if (error != NO_ERROR) {
        printf("Error reading SVM40 serial: %i\n", error);
    } else {
        printf("Serial Number: %s\n", serial_id);
    }

    struct svm40_version_information version_information;
    error = svm40_get_version(&version_information);
    if (error) {
        printf("Error reading SVM40 version: %i\n", error);
    } else {
        if (version_information.firmware_debug) {
            printf("Development firmware version: ");
        }
        printf("FW: %u.%u, HW: %u.%u, protocol: %u.%u\n",
               version_information.firmware_major,
               version_information.firmware_minor,
               version_information.hardware_major,
               version_information.hardware_minor,
               version_information.protocol_major,
               version_information.protocol_minor);
    }

    error = svm40_start_continuous_measurement();
    if (error) {
        printf("Error starting measurement: %i\n", error);
    }

    for (;;) {
        int16_t voc_index;
        int16_t relative_humidity;
        int16_t temperature;
        sensirion_sleep_usec(SVM40_MEASUREMENT_INTERVAL_USEC); /* wait 1s */
        error = svm40_read_measured_values_as_integers(
            &voc_index, &relative_humidity, &temperature);
        if (error) {
            printf("Error reading measurement values: %i\n", error);
        } else {
            printf("Measurement result\n");
            printf(" VOC index: %.1f\n", voc_index / 10.0f);
            printf(" Relative Humidity: %.2f%%RH\n",
                   relative_humidity / 100.0f);
            printf(" Temperature: %.2f°C\n", temperature / 200.0f);
        }
    }

    return 0;
}


svm40/svm40.h
/*
 * Copyright (c) 2020, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef SVM40_H
#define SVM40_H

#include "sensirion_arch_config.h"
#include "sensirion_common.h"
#include "sensirion_i2c.h"

#ifdef __cplusplus
extern "C" {
#endif

#define SVM40_I2C_ADDRESS 0x6a

#define SVM40_CMD_START_CONTINUOUS_MEASUREMENT 0x0010
#define SVM40_CMD_STOP_MEASUREMENT 0x0104
#define SVM40_CMD_READ_MEASURED_VALUES_AS_INTEGERS 0x03A6
#define SVM40_CMD_READ_MEASURED_VALUES_AS_INTEGERS_WITH_RAW_PARAMETERS 0x03B0
#define SVM40_CMD_GET_VERSION 0xD100
#define SVM40_CMD_GET_SERIAL_NUMBER 0xD033
#define SVM40_CMD_DEVICE_RESET 0xD304

#define SVM40_MEASUREMENT_INTERVAL_USEC 1000000
#define SVM40_MAX_SERIAL_LEN 26
#define SVM40_READ_DELAY 1000

/**
 * Check if SVM40 sensor is available
 *
 * @return  NO_ERROR on success, an error code otherwise
 */
int16_t svm40_probe(void);

struct svm40_version_information {

    /** Major part of the firmware version major.minor */
    uint8_t firmware_major;
    /** Minor part of the firmware version major.minor */
    uint8_t firmware_minor;
    /** Marks a development firmware version */
    bool firmware_debug;

    /** Major part of the hardware version major.minor */
    uint8_t hardware_major;
    /** Minor part of the hardware version major.minor */
    uint8_t hardware_minor;

    /** Major part of the protocol version major.minor */
    uint8_t protocol_major;
    /** Minor part of the protocol version major.minor */
    uint8_t protocol_minor;
};

/**
 * Read the hardware and firmware version information
 *
 * @param version_information   Struct to store the version information
 * @return  NO_ERROR on success, an error code otherwise
 */
int16_t
svm40_get_version(struct svm40_version_information* version_information);

/**
 * Get Serial Number.
 *
 * @param serial Memory where the serial number is written into.
 * 				 The string is guaranteed to be zero terminated.
 * 				 It must be at least SVM40_MAX_SERIAL_LEN long.
 * @return NO_ERROR on success, an error code otherwise
 */
int16_t svm40_get_serial(char* serial);

/**
 * Start a continuous measurement.
 *
 * @note This command is only available in idle mode.
 * @return  NO_ERROR on success, an error code otherwise
 */
int16_t svm40_start_continuous_measurement(void);

/**
 * Stop a running measurement.
 *
 * @note This command is only available in measurement mode.
 * @return  NO_ERROR on success, an error code otherwise
 */
int16_t svm40_stop_measurement(void);

/**
 * Read the current measured values.
 *
 * The firmware updates the measurement values every second. Polling data
 * faster will return the same values. The first measurement is available one
 * second after the start emasurement command is issued. Any readout prior to
 * this will return zero initialized values.
 *
 * @note Only availabe in measurement mode.
 * @param   voc_index           VOC algorithm output with a scaling value of 10.
 * @param   relative_humidity   Compensated ambient humidity in %RH with a
 *                              scaling factor of 100.
 * @param   temperature         Compensated ambient temperature in degree
 *                              celsius with a scaling factor of 200.
 * @return  NO_ERROR on success, an error code otherwise
 */
int16_t svm40_read_measured_values_as_integers(int16_t* voc_index,
                                               int16_t* relative_humidity,
                                               int16_t* temperature);

/**
 * Returns the new measurement results as integers along with the raw voc
 * ticks and uncompensated RH/T values.
 *
 * The firmware updates the measurement values every second. Polling data
 * faster will return the same values. The first measurement is available one
 * second after the start measurement command is issued. Any readout prior to
 * this will return zero initialized values.
 *
 * @note Only available in measurement mode.
 * @param   voc_index                       VOC algorithm output
 *                                          with a scaling value of 10.
 * @param   relative_humidity               Compensated ambient
 *                                          humidity in %RH with a scaling
 *                                          factor of 100.
 * @param   temperature                     Compensated ambient temperature
 *                                          in degree celsius with a scaling
 *                                          factor of 200.
 * @param   voc_ticks_raw                   Raw VOC output ticks as
 *                                          read from the SGP sensor.
 * @param   uncompensated_relative_humidity Uncompensated raw humidity in
 *                                          %RH as read from the SHT40 with a
 *                                          scaling factor of 100.
 * @param   uncompensated_temperature       Uncompensated raw temperature in
 *                                          degrees celsius as read from the
 *                                          SHT40 with a scaling of 200.
 * @return  NO_ERROR on success, an error code otherwise
 */
int16_t svm40_read_measured_values_as_integers_with_raw_params(
    int16_t* voc_index, int16_t* relative_humidity, int16_t* temperature,
    uint16_t* voc_ticks_raw, int16_t* uncompensated_relative_humidity_raw,
    int16_t* uncompensated_temperature);

/**
 * Return the driver version
 *
 * @return  Driver version string
 */
const char* svm40_get_driver_version(void);

#ifdef __cplusplus
}
#endif

#endif /* SVM40_H */
